From 575316c4706e5600f3935e8245027f7de6222656 Mon Sep 17 00:00:00 2001
From: Serge Hallyn <serge.hallyn@ubuntu.com>
Date: Tue, 19 Apr 2016 13:30:29 -0500
Subject: [PATCH 1/2] access: allow rx to controller directories themselves

Signed-off-by: Serge Hallyn <serge.hallyn@ubuntu.com>
---
 bindings.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/bindings.c b/bindings.c
index 97fc397..3f32816 100644
--- a/bindings.c
+++ b/bindings.c
@@ -1791,8 +1791,10 @@ int cg_access(const char *path, int mode)
 	if (!controller)
 		return -EIO;
 	cgroup = find_cgroup_in_path(path);
-	if (!cgroup)
-		return -EINVAL;
+	if (!cgroup) {
+		// access("/sys/fs/cgroup/systemd", mode) - rx allowed, w not
+		return mode & W_OK == 0 ? 0 : -EACCES;
+	}
 
 	get_cgdir_and_path(cgroup, &cgdir, &last);
 	if (!last) {
-- 
2.7.4

