From 2d9cd3b3655319710e1baafad9453ef69e75e950 Mon Sep 17 00:00:00 2001
From: Christian Brauner <christian.brauner@ubuntu.com>
Date: Mon, 12 Jun 2017 14:45:44 +0200
Subject: [PATCH 06/11] bindings: calculate btime correctly

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>
---
 bindings.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/bindings.c b/bindings.c
index 3aaf5f9..f3aefa1 100644
--- a/bindings.c
+++ b/bindings.c
@@ -3606,7 +3606,7 @@ static uint64_t get_reaper_btime(pid)
 
 	uptime = (uint64_t)time(NULL) - (uint64_t)sys.uptime;
 	procstart = get_reaper_start_time_in_sec(pid);
-	return uptime - procstart;
+	return uptime + procstart;
 }
 
 #define CPUALL_MAX_SIZE (BUF_RESERVE_SIZE / 2)
-- 
2.13.2

