From 3b6c5c19738e9c259f51ab35159d6b33bf5b6029 Mon Sep 17 00:00:00 2001
From: Serge Hallyn <serge@hallyn.com>
Date: Fri, 17 Jun 2016 02:27:45 -0500
Subject: [PATCH] under_systemd_user_slice: work right when init is in '/'

Signed-off-by: Serge Hallyn <serge@hallyn.com>
---
 pam/pam_cgfs.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/pam/pam_cgfs.c b/pam/pam_cgfs.c
index 3fd11d8..08f0694 100644
--- a/pam/pam_cgfs.c
+++ b/pam/pam_cgfs.c
@@ -477,6 +477,8 @@ static bool under_systemd_user_slice(struct controller *c, uid_t uid)
 	if (strncmp(c->cur_path, c->init_path, initlen) != 0)
 		return false;
 	snprintf(buf, 100, "/user.slice/user-%d.slice/", (int)uid);
+	if (initlen == 1)
+		initlen = 0; // skip the '/'
 	return strncmp(c->cur_path + initlen, buf, strlen(buf)) == 0;
 }
 
